File: leetcode/hashMap/160.cpp
#include <algorithm>
#include <functional> //包含std::greater 和 std::less
#include <iostream>
#include <limits.h>
#include <math.h>
#include <queue>
#include <stack>
#include <unordered_set>
#include <vector>

using namespace std;
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution {
public:
  ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
    unordered_set<ListNode *> umap;
    ListNode *cur = headA;
    while (nullptr != cur) {
      umap.insert(cur);
      cur = cur->next;
    }
    cur = headB;
    while (nullptr != cur) {
      if (umap.end() != umap.find(cur)) {
        return cur;
      }
      cur = cur->next;
    }
    return nullptr;
  }
};
int main() {
  // 示例二叉树
  Solution solution;
  vector<int> nums = {7, 1, 5, 3, 6, 4};
  ListNode *headA = new ListNode(), *headB = new ListNode(),
           *headC = new ListNode();
  headA->val = 4;
  headA->next = new ListNode(1);
  ListNode *common = new ListNode(8);
  headA->next->next = common;
  common->next = new ListNode(4);
  common->next->next = new ListNode(5);
  headB->val = 5;
  headB->next = new ListNode(6);
  headB->next->next = new ListNode(1);
  headB->next->next->next = common;
  solution.getIntersectionNode(headA, headB);
  //   std::cout << "result: " << result << std::endl;
}

File: leetcode/hashMap/160.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next: Optional["ListNode"] = None


class Solution:
    def getIntersectionNode(
        self, headA: ListNode, headB: ListNode
    ) -> Optional[ListNode]:
        visited = set()
        cur = headA
        while cur:
            visited.add(cur)
            cur = cur.next
        cur = headB
        while cur:
            if cur in visited:
                return cur
            cur = cur.next
        return None


if __name__ == "__main__":
    # 构造测试用例
    solution = Solution()
    # 公共部分
    common = ListNode(8)
    common.next = ListNode(4)
    common.next.next = ListNode(5)
    # headA: 4 -> 1 -> 8 -> 4 -> 5
    headA = ListNode(4)
    headA.next = ListNode(1)
    headA.next.next = common
    # headB: 5 -> 6 -> 1 -> 8 -> 4 -> 5
    headB = ListNode(5)
    headB.next = ListNode(6)
    headB.next.next = ListNode(1)
    headB.next.next.next = common
    result = solution.getIntersectionNode(headA, headB)
    print(f"result = {result.val if result else None}")
